### PR TITLE
Add model metadata to system index metadata

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
For users to specify a model in their knn_vector, they had also needed to specify the engine, spacetype and dimension:
```
PUT /my-knn-index
{
    "mappings": {
       "properties": {
           "my-knn-field": {
               "type": "knn_vector",
               "dimension": 16
               "model": {
                   "model_id": "my-model-id",
                   "engine": "faiss",
                   "space_type": "l2"
               }
           }
       }
   }
}
```

While this information is present in the model index, blocking calls could not be made to retrieve it in the field mapper, and the code could not be structured in a non-blocking way. So, the original implementation (#68) required redundant info from the customer.

This PR removes the redundancy by adding model metadata to the model system index so that blocking/non-blocking calls are not necessary. With this change, a user can specify a mapping as such:
```
PUT /my-knn-index
{
    "mappings": {
       "properties": {
           "my-knn-field": {
               "type": "knn_vector",
               "model_id": "my_model_id"
           }
       }
   }
}
```

To accomplish this, this PR creates a master transport operation that stores model metadata in a map in the model index's custom metadata. On model put/delete, this metadata gets updated. To access the metadata, a user can call `getMetadata(modelId)` from the modelDao.

Note - this PR builds on top of #88
 
### Issues Resolved
#27 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 1]
```diff
@@ -215,13 +232,27 @@ public void put(Model model, ActionListener<IndexResponse> listener) throws IOEx
...
+                create(ActionListener.wrap(createIndexResponse -> indexRequestBuilder.execute(putMetadataListener),
                         listener::onFailure));
                 return;
             }
 
-            indexRequestBuilder.execute(listener);
+            indexRequestBuilder.execute(putMetadataListener);
+        }
+
+        private ActionListener<IndexResponse> getModelIndexListener(ModelMetadata modelMetadata,
```
[the remark this answers is not in the document]
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:249]: I see. Yes I can refactor this. I think because it is a function and not a class we can call it `getUpdateModelMetadataListener`

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 2]
```diff
@@ -244,9 +275,31 @@ public Model get(String modelId) throws ExecutionException, InterruptedException
...
                         + modelId + "\".");
             }
+            ModelMetadata modelMetadata = new ModelMetadata(KNNEngine.getEngine((String) engine),
+                    SpaceType.getSpace((String) space), (Integer) dimension);
+            return new Model(modelMetadata, Base64.getDecoder().decode((String) blob));
+        }
+
+        @Override
+        public ModelMetadata getMetadata(String modelId) {
+            IndexMetadata indexMetadata = clusterService.state().metadata().index(MODEL_INDEX_NAME);
```
[the remark this answers is not in the document]
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:285]: This gets the cluster state on the local node:
1. [clusterService.state()](https://github.com/jmazanec15/k-NN-1/blob/system-index-metadata/src/main/java/org/opensearch/knn/indices/ModelDao.java#L285)
2. [clusterApplierService.state()](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/service/ClusterService.java#L150)
3. [state.get()](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L202) - which gets it from atomic ref..

File: src/main/java/org/opensearch/knn/indices/ModelMetadata.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.opensearch.knn.index.KNNVectorFieldMapper.MAX_DIMENSION;
+
+public class ModelMetadata implements Writeable {
+
+    public static final String MODEL_METADATA_FIELD = "knn-models"; //TODO: Move this to constants
+    private static final String DELIMITER = ","; // TODO: This doesnt seem to belong there
```
[the remark this answers is not in the document]
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelMetadata.java]: fixed

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java [hunk 4]
```diff
@@ -266,21 +268,32 @@ public KNNVectorFieldMapper build(BuilderContext context) {
     }
 
     public static class TypeParser implements Mapper.TypeParser {
+
+        // Use a supplier here because in {@link org.opensearch.knn.KNNPlugin#getMappers()} the ModelDao has not yet
+        // been initialized
+        private Supplier<ModelDao> modelDaoSupplier;
+
+        public TypeParser(Supplier<ModelDao> modelDaoSupplier) {
+            super();
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:277]: Looks like TypeParser only implements not extends. Do we need super?

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java [hunk 5]
```diff
@@ -226,17 +224,21 @@ public KNNVectorFieldMapper build(BuilderContext context) {
                         knnMethodContext);
             }
 
-            ModelContext modelContext = this.modelContext.get();
-            if (modelContext != null) {
+            String modelId = this.modelId.get();
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:227]: May be different name to avoid collision with instance name?

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java [hunk 6]
```diff
@@ -226,17 +224,21 @@ public KNNVectorFieldMapper build(BuilderContext context) {
...
+                        new KNNVectorFieldType(buildFullName(context), meta.getValue(), modelMetadata.getDimension()),
                         multiFieldsBuilder.build(this, context),
                         copyTo.build(),
                         ignoreMalformed(context),
                         stored.get(),
                         hasDocValues.get(),
-                        modelContext);
+                        modelDao,
+                        modelId,
+                        modelMetadata);
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:241]: This can be extracted from modelId and ModelDAO which are passed already. Do we still need to pass?

File: src/main/java/org/opensearch/knn/indices/ModelDao.java [hunk 7]
```diff
@@ -76,7 +81,7 @@
      * @param model Model to be indexed
      * @param listener  handles index response
      */
-    void put(String modelId, Model model, ActionListener<IndexResponse> listener) throws IOException;
+    void put(String modelId, Model model, ActionListener<AcknowledgedResponse> listener) throws IOException;
```
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/indices/ModelDao.java:84]: Listener doc is not updated

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java (hunk 5 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:227]: Sure, I can update.

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java (hunk 6 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:241]: There is a reason, but it is somewhat convoluted. 
1. In 229 we have to get modelMetadata so that we can retrieve the dimension to build the MappedFieldType in 233
2. In 233 we need to build the MappedFieldType here because in 357, KNNVectorFieldMapper constructor takes a MappedFieldType as a construction arg. If we wanted to build MappedFieldType in 600, we would need to call `modelDao.getMetadata(modelId).getDimension()` in KNNVectorFieldType's constructor. Then, we would need to call `modelDao.getMetadata(modelId)` again later in ModelFieldMapper's constructor.

So, tldr, the purpose of this is to prevent duplicate calls to the modelDao.

So, why do we need to pass modelDao then to the ModelFieldMapper's constructor? 
1. In 452, getMergeBuilder passes the modelDao to the builders constructor.
2. In 169, we have ModelDao be a parameter (as opposed to just getting the singleton) to make it easier to test.

I welcome any opinions on any better approaches.

File: src/main/java/org/opensearch/knn/indices/ModelDao.java (hunk 7 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/indices/ModelDao.java:84]: Will update.

File: src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java (hunk 4 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java:277]: Ah good catch. No we do not

**Review by vamshin — APPROVED** [review]
LGTM! Thanks
